File: Binary Search/search2Dmatrix.cpp
#include <bits/stdc++.h>

using namespace std;

bool searchMatrix(vector<vector<int>> &matrix, int target)
{

    int lowRow = 0;
    int highRow = matrix.size() - 1;

    while (lowRow < highRow)
    {
        int mid = lowRow + (highRow - lowRow) / 2;

        if (matrix[mid][0] == target)
        {
            return true;
        }

        if (matrix[mid][0] < target && target < matrix[mid + 1][0])
        {
            lowRow = mid;
            break;
        }

        if (matrix[mid][0] < target)
        {
            lowRow = mid + 1;
        }

        else
        {
            highRow = mid - 1;
        }
    }

    int low = 0;
    int high = matrix[0].size() - 1;

    while (low <= high)
    {
        int mid = low + (high - low) / 2;

        if (matrix[lowRow][mid] == target)
        {
            return true;
        }

        if (matrix[lowRow][mid] < target)
        {
            low = mid + 1;
        }

        else
        {
            high = mid - 1;
        }
    }

    return false;
}

File: Binary Search/Search_2D_Matrix.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        top = 0
        bottom = len(matrix) - 1

        while top <= bottom:
            mid = top + (bottom - top) // 2

            if target < matrix[mid][0]:
                bottom = mid - 1
            elif target > matrix[mid][-1]:
                top = mid + 1
            else:
                break

        if not (top <= bottom):
            return False

        row = top + (bottom - top) // 2

        l = 0
        r = len(matrix[row]) - 1

        while l <= r:
            mid = l + (r - l) // 2

            if matrix[row][mid] < target:
                l = mid + 1

            elif matrix[row][mid] > target:
                r = mid - 1

            else:
                return True

        return False
